### ISSUE TITLE
Marten Command Failure: CREATE OR REPLACE FUNCTION public.mt_immutable_timestamp(value text)

### ISSUE THREAD

**Issue by stvoidmain** [JasperFx/marten#864]
"Marten Command Failure:\nCREATE OR REPLACE FUNCTION public.mt_immutable_timestamp(value text) RETURNS timestamp with time zone LANGUAGE sql IMMUTABLE AS $function$\n    select value::timestamptz\n$function$;\r\n\r\n\n\n3F000: schema «public» does not exists"

When i configure marten with custom schema and in the database there is no "public" schema throws that error, I found that https://github.com/JasperFx/marten/issues/837 was closed after posting, sorry, but, in any case, its not the exact same problem, is related, tho.
It should never look for it in public schema, why we assume there is always one? It's just a convention, my Postgre databases almost never have a public schema i keep droping it after creating database 99% of the times, and, in any case, it would be better to fix the Add of SystemFunctions to take DabaseSchemaName into account, or is there any reason that i'm missing to have all in public schmea and make all marten user to have a default schema by martens convention?

**Comment by jeremydmiller** [comment]
@stvoidmain You're the very, very first person to have an issue with this. Marten is good enough to auto-create schemas on the fly in its development mode. I believe that the fix for #837 made it honor the `StoreOptions.DatabaseSchemaName`, or have you found some pathway in the code where that's not true?

**Comment by stvoidmain** [comment]
First of all, marten is great, i love it, amazing work ;)

I found one, if you configure your store like this:
```
    store = DocumentStore.For(_ =>
    {
        _.DatabaseSchemaName = "documents";
        var connStr = Configuration.GetConnectionString("DefaultConnection");
        _.Connection(connStr);
    });
```
and you have no 'public' schema in your db, fails with the error in the first post.
"schema «public» does not exist"...

**Comment by stvoidmain** [comment]
The patch for https://github.com/JasperFx/marten/issues/837 did exactly the opposite, it changed the use of DatabaseSchemaName for a hard coded call to "public."...

**Comment by jeremydmiller** [comment]
Ugh, yeah, I see what you mean. I missed it reviewing the PR. Interested in fixing this again, but buy putting that function in your main Marten database scheme instead?

**Comment by stvoidmain** [comment]
Sure thing, want me to do a PR?

**Comment by mysticmind** [comment]
@stvoidmain and @jeremydmiller the function always gets created in public immaterial of database scheme and hence had fixed the logic to use public. 

I am looking into it and found a deeper underlying issue which I will discuss with @jeremydmiller  and will get this going, I am on it.

**Comment by mysticmind** [comment]
As a first step, I have reverted the incorrect fix and sent a PR #865.

The underlying problem is with the `StoreOptions` class constructor as below:
```
public StoreOptions()
 {
     Events = new EventGraph(this);
     Schema = new MartenRegistry();
     Transforms = new Transforms.Transforms(this);
     Storage = new StorageFeatures(this);
}
```
During the constructor call, database schema name value is still `public` and the constructor logic will always instanciate with SystemFunctions etc under public schema. Hence immaterial of where we set a schema name or not, the functions are getting created under `public` schema.

The only way I see a fix is to pass the database schema name as a parameter to the `StoreOptions` constructor rather than via the Action<T>. And also disallow, the setter for `DatabaseSchemeName` (this would be a breaking change).
